File: 04-Design-Principles/src/go/weapon.go
package main

import "fmt"

// Go naturally favors composition — interfaces are implicit, embedding is composition

type WeaponBehavior interface {
	Attack(name string)
}

type SwordWeapon struct{}
func (SwordWeapon) Attack(name string) { fmt.Printf("%s swings a sword!\n", name) }

type BowWeapon struct{}
func (BowWeapon) Attack(name string) { fmt.Printf("%s shoots an arrow!\n", name) }

type MagicWeapon struct{}
func (MagicWeapon) Attack(name string) { fmt.Printf("%s casts a fireball!\n", name) }

File: 04-Design-Principles/src/rust/weapon.rs
// Rust naturally favors composition over inheritance (no inheritance at all!)
// Traits = behaviors, structs compose via fields

pub trait WeaponBehavior {
    fn attack(&self, name: &str);
}

pub struct SwordWeapon;
impl WeaponBehavior for SwordWeapon {
    fn attack(&self, name: &str) { println!("{} swings a sword!", name); }
}

pub struct BowWeapon;
impl WeaponBehavior for BowWeapon {
    fn attack(&self, name: &str) { println!("{} shoots an arrow!", name); }
}

pub struct MagicWeapon;
impl WeaponBehavior for MagicWeapon {
    fn attack(&self, name: &str) { println!("{} casts a fireball!", name); }
}
